File: NCPC/2015/ncpc2015/heropower/submissions/time_limit_exceeded/Bj0rnStepHalfMilliseconds.cpp
#include <cstdio>
#include <climits>
#include <algorithm>

using namespace std;

static int END = INT_MAX / 2;

struct Phrase
{
	int *start, *end;
};

int *fromPhraseOptimals;
Phrase *phraseArray;
int optimalFromPhrase(Phrase *phrases);

int optimizeFromPhrase(Phrase *phrases)
{
	int maxBonus = -1;
	int doubleSP = 0;
	int *noteAt = phrases->start;
	int *noteAfter = noteAt;
	Phrase *activePhrase = phrases;
	Phrase *phraseAfter = activePhrase + 1;

	// Slow loop - step half a millisecond at a time.
	for (int doublePosition = *phrases->start * 2; *noteAfter < END;
		 doublePosition++,
				 doubleSP += doublePosition > *activePhrase->start * 2 && doublePosition <= *activePhrase->end * 2 ? 1 : 0,
				 activePhrase += doublePosition == *activePhrase->end * 2 ? 1 : 0)
	{
		if (doublePosition > *noteAt * 2)
			noteAt++;
		if (doublePosition + doubleSP > *noteAfter * 2)
			noteAfter++;
		if (doublePosition + doubleSP > *phraseAfter->start * 2)
			phraseAfter++;
		int bonusOnActivation = noteAfter - noteAt + optimalFromPhrase(phraseAfter);
		maxBonus = bonusOnActivation > maxBonus ? bonusOnActivation : maxBonus;
	}
	return maxBonus;
}


int optimalFromPhrase(Phrase *phrases)
{
	int offset = phrases - phraseArray;
	if (fromPhraseOptimals[offset] == -1)
		fromPhraseOptimals[offset] = optimizeFromPhrase(phrases);
	return fromPhraseOptimals[offset];
}

int main()
{
	int n, p;
	scanf("%d%d", &n, &p);
	int *notes = new int[n + 1];
	for (int i = 0; i < n; ++i)
		scanf("%d", &notes[i]);
	notes[n] = END;

	Phrase *phrases = new Phrase[p + 1];
	for (int i = 0; i < p; ++i)
	{
		int start, end;
		scanf("%d%d", &start, &end);
		phrases[i].start = find(notes, &notes[n], start);
		phrases[i].end   = find(notes, &notes[n], end  );
	}
	phrases[p] = { &notes[n], &notes[n] };

	fromPhraseOptimals = new int[p + 1];
	for (int i = 0; i < p; ++i)
		fromPhraseOptimals[i] = -1;
	fromPhraseOptimals[p] = 0;
	phraseArray = phrases;

	printf("%d\n", n + optimalFromPhrase(&phrases[0]));

	return 0;
}

File: NCPC/2015/ncpc2015/heropower/submissions/time_limit_exceeded/Bj0rnStepHalfMilliseconds.java
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.io.OutputStream;

public class Bj0rnStepHalfMilliseconds {
    static class Phrase
    {
        int start, end;
        Phrase(int start, int end) { this.start = start; this.end = end; }
    }

    public static void main(String[] args) {
        Kattio io = new Kattio(System.in);
        int n = io.getInt(), p = io.getInt();
        int[] notes = new int[n];
        for (int i = 0; i < n; i++)
            notes[i] = io.getInt();

        Phrase[] phrases = new Phrase[p];
        int ni = 0;
        for (int i = 0; i < p; i++)
        {
            int position = io.getInt();
            for (; notes[ni] < position; ni++);
            int start = ni;

            position = io.getInt();
            for (; notes[ni] < position; ni++);
            int end = ni;

            phrases[i] = new Phrase(start, end);
        }

        System.out.println(n + new Optimizer(notes, phrases).bonusFromPhrase(0));
    }

    static class Optimizer
    {
        int[] notes;
        Phrase[] phrases;
        int[] fromPhraseBonus;

        Optimizer(int[] notes, Phrase[] phrases)
        {
            this.notes = notes;
            this.phrases = phrases;
            fromPhraseBonus = new int[phrases.length + 1];
            for (int i = 0; i < phrases.length; i++)
                fromPhraseBonus[i] = -1;
            fromPhraseBonus[phrases.length] = 0;
        }

        int bonusFromPhrase(int phrase)
        {
            int bonus = fromPhraseBonus[phrase];
            if (bonus != -1)
                return bonus;
            else
            {
                int doubleSP = 0;
                int noteAt = phrases[phrase].start;
                int noteAfter = noteAt;
                int activePhrase = phrase;
                int phraseAfter = activePhrase + 1;

                // Slow loop - step half a millisecond at a time.
                for (int doublePosition = notes[phrases[phrase].start] * 2; doublePosition <= notes[notes.length-1] * 2;
                     doublePosition++,
                         doubleSP += activePhrase < phrases.length && doublePosition > notes[phrases[activePhrase].start] * 2 && doublePosition <= notes[phrases[activePhrase].end] * 2 ? 1 : 0,
                         activePhrase += activePhrase < phrases.length && doublePosition == notes[phrases[activePhrase].end] * 2 ? 1 : 0)
                {
                    if (doublePosition > notes[noteAt] * 2)
                        noteAt++;
                    if (noteAfter < notes.length && doublePosition + doubleSP > notes[noteAfter] * 2)
                        noteAfter++;
                    if (phraseAfter < phrases.length && doublePosition + doubleSP > notes[phrases[phraseAfter].start] * 2)
                        phraseAfter++;
                    int bonusOnActivation = noteAfter - noteAt + bonusFromPhrase(phraseAfter);
                    bonus = bonusOnActivation > bonus ? bonusOnActivation : bonus;
                }
                fromPhraseBonus[phrase] = bonus;
                return bonus;
            }
        }
    }
}

/** Simple yet moderately fast I/O routines.
 *
 * Example usage:
 *
 * Kattio io = new Kattio(System.in, System.out);
 *
 * while (io.hasMoreTokens()) {
 *    int n = io.getInt();
 *    double d = io.getDouble();
 *    double ans = d*n;
 *
 *    io.println("Answer: " + ans);
 * }
 *
 * io.close();
 *
 *
 * Some notes:
 *
 * - When done, you should always do io.close() or io.flush() on the
 *   Kattio-instance, otherwise, you may lose output.
 *
 * - The getInt(), getDouble(), and getLong() methods will throw an
 *   exception if there is no more data in the input, so it is generally
 *   a good idea to use hasMoreTokens() to check for end-of-file.
 *
 * @author: Kattis
 */

class Kattio extends PrintWriter {
    public Kattio(InputStream i) {
        super(new BufferedOutputStream(System.out));
        r = new BufferedReader(new InputStreamReader(i));
    }
    public Kattio(InputStream i, OutputStream o) {
        super(new BufferedOutputStream(o));
        r = new BufferedReader(new InputStreamReader(i));
    }

    public boolean hasMoreTokens() {
        return peekToken() != null;
    }

    public int getInt() {
        return Integer.parseInt(nextToken());
    }

    public double getDouble() {
        return Double.parseDouble(nextToken());
    }

    public long getLong() {
        return Long.parseLong(nextToken());
    }

    public String getWord() {
        return nextToken();
    }



    private BufferedReader r;
    private String line;
    private StringTokenizer st;
    private String token;

    private String peekToken() {
        if (token == null)
            try {
                while (st == null || !st.hasMoreTokens()) {
                    line = r.readLine();
                    if (line == null) return null;
                    st = new StringTokenizer(line);
                }
                token = st.nextToken();
            } catch (IOException e) { }
        return token;
    }

    private String nextToken() {
        String ans = peekToken();
        token = null;
        return ans;
    }
}
